### PR TITLE
[flang][OpenMP] Update handling of DEPEND clause

### PR DESCRIPTION
**Pull request by kparzysz**
Parse the locator list in OmpDependClause as an OmpObjectList (instead of a list of Designators). When a common block appears in the locator list, show an informative message.
Implement resolving symbols in DependSinkVec in a dedicated visitor instead of having a visitor for OmpDependClause.
Resolve unresolved names common blocks in OmpObjectList.

Minor changes to the code organization:
- rename OmpDependenceType to OmpTaskDependenceType (to follow 5.2 terminology),
- rename Depend::WithLocators to Depend::DepType,
- add comments with more detailed spec references to parse-tree.h.

### CODE REVIEW

**Comment by kparzysz** [comment]
Background: The OpenMP spec states that a common block name cannot appear in the locator list, so I wanted to emit a diagnostic for this.  That wasn't hard, but I was seeing an additional unexpected error about an unresolved symbol in the common block.  I tracked it down to the symbol resolution never visiting common blocks in OmpObjectList.

The reason it happened is that `ResolveNamesVisitor` (which is the main driver of the AST walk) didn't have a Pre/Post handler for an individual `parser::Name` (which is used to represent a common block name).  Instead, each potential owner of such a name would deal with it separately.
Moreover, in the `OmpAttributeVisitor` we had a Pre handler for OmpDependClause, which only visited the sink vector and returned "false".  This return value would prevent recursing into that object any further.  To allow visiting the OmpObjectList, without interfering with the clause traveral, I extracted the visit of the sink vector into its own handler.  Finally, I added a post-visitor for OmpObjectList to resolve any remaining common blocks. [1]

[1] Named common blocks are valid list items.  There are some clauses that disallow it.  I don't think we handled all of that correctly in all cases, and this should get us closer to that.

**Comment by kparzysz** [comment]
Ping.  This is the beginning of the PR chain for changes related to DEPEND.

File: flang/lib/Semantics/check-omp-structure.cpp [hunk 1]
```diff
@@ -3283,15 +3283,21 @@ void OmpStructureChecker::Enter(const parser::OmpClause::Depend &x) {
...
-                std::get_if<common::Indirection<parser::ArrayElement>>(
-                    &dataRef->u)}) {
-          CheckArraySection(arr->value(), GetLastName(*dataRef),
-              llvm::omp::Clause::OMPC_depend);
+    for (const auto &object : std::get<parser::OmpObjectList>(inOut->t).v) {
+      if (const auto *name{std::get_if<parser::Name>(&object.u)}) {
+        context_.Say(GetContext().clauseSource,
+            "Common block name ('%s') cannot appear in a DEPEND "
+            "clause"_err_en_US,
+            name->ToString());
```
Comment (kiranchandramohan) [inline flang/lib/Semantics/check-omp-structure.cpp:3291]: Is ToString required?

File: flang/include/flang/Parser/parse-tree.h [hunk 2]
```diff
@@ -3439,6 +3439,18 @@ struct OmpObject {
 
 WRAPPER_CLASS(OmpObjectList, std::list<OmpObject>);
 
+// Ref: [4.5:169-170], [5.0:254-256], [5.1:287-289], [5.2:321]
+//
+// task-dependence-type -> // "dependence-type" in 5.1 and before
+//    IN | OUT | INOUT |        // since 4.5
+//    SOURCE | SINK |           // since 4.5, until 5.1
+//    MUTEXINOUTSET | DEPOBJ |  // since 5.0
+//    INOUTSET                  // since 5.2
```
Comment (kiranchandramohan) [inline flang/include/flang/Parser/parse-tree.h:3448]: Are MUTEXINOUTSET, DEPOBJ, INOUTSET going to come in later patches?

File: flang/lib/Lower/OpenMP/ClauseProcessor.cpp [hunk 3]
```diff
@@ -798,11 +798,11 @@ bool ClauseProcessor::processDepend(mlir::omp::DependClauseOps &result) const {
   return findRepeatableClause<omp::clause::Depend>(
       [&](const omp::clause::Depend &clause, const parser::CharBlock &) {
         using Depend = omp::clause::Depend;
-        assert(std::holds_alternative<Depend::WithLocators>(clause.u) &&
-               "Only the modern form is handled at the moment");
-        auto &modern = std::get<Depend::WithLocators>(clause.u);
-        auto kind = std::get<Depend::TaskDependenceType>(modern.t);
-        auto &objects = std::get<omp::ObjectList>(modern.t);
+        assert(std::holds_alternative<Depend::DepType>(clause.u) &&
+               "Only the form with depenence type is handled at the moment");
```
Comment (kiranchandramohan) [inline flang/lib/Lower/OpenMP/ClauseProcessor.cpp:802]:
```suggestion
               "Only the form with dependence type is handled at the moment");
```

**Review by kiranchandramohan — COMMENTED** [review]
Sorry missed this. Looks OK. A few minor comments.

File: flang/lib/Semantics/check-omp-structure.cpp (hunk 1 above)
Comment (kparzysz) [reply flang/lib/Semantics/check-omp-structure.cpp]: Not necessary.  Removed.

File: flang/include/flang/Parser/parse-tree.h (hunk 2 above)
Comment (kiranchandramohan) [reply flang/include/flang/Parser/parse-tree.h:3448]: OK

**Review by kiranchandramohan — APPROVED** [review]
LGTM.

**Pull request merged**